File: client_app/lib/models/client_order.dart
import 'package:client_app/models/dish.dart';
import 'package:client_app/models/restaurant.dart';
import 'package:client_app/models/user.dart';
import 'package:flutter/material.dart';

class ClientOrder {
  int id;
  int clientId;
  int restaurantId;
  int totalPrice;
  String status;
  String type;
  DateTime createdAt;
  List<Dish> itemList=[];

  ClientOrder({
    this.id, 
    @required this.clientId,
    @required this.restaurantId, 
    this.type,
    this.status,
    this.createdAt,
    this.totalPrice=0, 
    this.itemList});

  void addDish(Dish dish) {
    itemList.add(dish);
    totalPrice+=dish.price;
  }

  void removeDish(Dish dish){
    for(int i=0; i<itemList.length; i++){
      if(itemList[i].id==dish.id){
        itemList.removeAt(i);
        totalPrice-=dish.price;
        break;
      }
    }
  }

  Map<String, dynamic> toMap() {
    return {
      'clientId': clientId,
      'restaurantId': restaurantId,
      'orderItems': itemList.map((dish) => dish.id).toList()
    };
  }

  static ClientOrder fromMap(Map<String, dynamic> orderMap){
    return ClientOrder(
      id:orderMap['id'],
      clientId: orderMap['client'], 
      restaurantId: orderMap['restaurant']['id'],
      type: orderMap['type'],
      status: orderMap['status'],
      createdAt: DateTime.parse(orderMap['createdAt']),
      totalPrice: orderMap['totalPrice'], 
      itemList: (orderMap['dishList'] as List).map((dishMap)=> Dish.fromMap(dishMap)).toList());
  }
}

File: order-mgmt/src/main/kotlin/com/order/ordermgmt/models/ClientOrder.kt
package com.order.ordermgmt.models

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

@Entity
class ClientOrder(
        @ManyToOne
        @JsonIgnore
        val client:Client,
        val type: OrderType,
        var status: OrderStatus,
        val createdAt: LocalDateTime,
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name="restaurant_id")
        val restaurant: Restaurant
):BaseEntity() {
    var totalPrice:Int=0
    @ManyToMany()
    @JoinTable(name="client_order_dish",
            joinColumns = [(JoinColumn(name="order_id", referencedColumnName = "id"))],
            inverseJoinColumns = [(JoinColumn(name = "dish_id",referencedColumnName = "id"))])
    @JsonIgnoreProperties("restaurant","orderList")
    val dishList= mutableListOf<Dish>()

    fun addDish(dish: Dish){
//        TODO: check for user sold
        dishList.add(dish)
        totalPrice+=dish.price
        client.sold-=dish.price
    }

    //just for seeding
    fun addDishList(list:MutableList<Dish>):ClientOrder{
        list.forEach {
            dishList.add(it)
            totalPrice+=it.price
        }
        return this
    }
}